### PR TITLE
[Harmony] Adding get validators and get delegations #733

### PR DESCRIPTION
**Pull request by coolcottontail**
Moved gupadhyaya 's old PR (https://github.com/trustwallet/blockatlas/pull/733) here

### CODE REVIEW

File: platform/harmony/stake.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	}
+}
+
+func normalizeValidator(v Validator, apr float64) (validator blockatlas.Validator, ok bool) {
+	return blockatlas.Validator{
+		Status:  true,
+		ID:      v.Address,
+		Details: getDetails(apr),
+	}, true
+}
```
Comment (prazd) [inline platform/harmony/stake.go]: maybe it's better to remove return of bool from here?
the `true` is constantly returning

Comment (EnoRage) [reply platform/harmony/stake.go]: We need to remove it

Comment (coolcottontail) [reply platform/harmony/stake.go]: removed.

**Comment by EnoRage** [comment]
Please, reopen the PR and fix the functional tests.

Issue 1: http://localhost:8420/v2/harmony/staking/validators
```
{
    "code": 500,
    "error": {
        "error": "unexpected end of JSON input",
        "meta": {},
        "type": "Platform Unmarshal Error"
    }
}
```

Issue 2:  http://localhost:8420/v2/harmony/staking/delegations/one1e4mr7tp0a76wnhv9xd0wzentdjnjnsh3fwzgfv

```
{
    "code": 500,
    "error": {
        "error": "RPC Call error: Unable to fetch delegations list",
        "meta": {
            "error_code": -32601,
            "error_message": "The method hmy_getDelegationsByDelegator does not exist/is not available",
            "method": "hmy_getDelegationsByDelegator"
        }
    }
}
```

**Comment by coolcottontail** [comment]
@EnoRage Can you tell me how to run your test cases ? 
The issue is that the endpoint has been changed to https://api.s0.os.hmny.io/.

**Comment by EnoRage** [comment]
@coolcottontail
in root of project 
`make test` for Unit tests
`make functional` for Functional tests (you will need all endpoints for config.yml  - not neccessary for your case)
`make integration` for integration tests
Can you please just finish previous issues, new response with updated endpoints (your last commit in branch) is:
Endpoint:
`http://localhost:8420/v2/harmony/staking/delegations/one1e4mr7tp0a76wnhv9xd0wzentdjnjnsh3fwzgfv`
```
{
    "code": 500,
    "error": {
        "error": "json: cannot unmarshal array into Go value of type harmony.Delegations: json-rpc GetObject Unmarshal error: Unable to fetch delegations list",
        "meta": {
            "obj": {
                "result": null
            },
            "string": "[]"
        }
    }
}
```

endpoint:
`http://localhost:8420/v2/harmony/staking/validators`

```
{
    "code": 500,
    "error": {
        "error": "unexpected end of JSON input",
        "meta": {
            "method": "GET",
            "path": "hmy_getAllValidatorAddresses"
        },
        "type": "Platform Unmarshal Error"
    }
}
```

**Comment by coolcottontail** [comment]
@EnoRage 
I can only run make test without any issue.  
make integration and make functional all failed even if I use the origin/master branch without any modification. Do I need to source any script before running tests ?

which test case will trigger this error ?  I am trying to duplicate and find out the reason... 

http://localhost:8420/v2/harmony/staking/delegations/one1e4mr7tp0a76wnhv9xd0wzentdjnjnsh3fwzgfv

{
    "code": 500,
    "error": {
        "error": "json: cannot unmarshal array into Go value of type harmony.Delegations: json-rpc GetObject Unmarshal error: Unable to fetch delegations list",
        "meta": {
            "obj": {
                "result": null
            },
            "string": "[]"
        }
    }
}


make integration failed:

  --- FAIL: Test/test_batch_with_invalid_coin (0.00s)
        domains.go:117: 
            	Error Trace:	domains.go:117
            	Error:      	Expected nil, but got: &errors.Error{Err:(*errors.errorString)(0xc000c3a760), Type:0x0, meta:map[string]interface {}{"coins":[]uint64{0x2c, 0x3c}, "name":"vitalik.eth"}, stack:[]string(nil)}
            	Test:       	Test/test_batch_with_invalid_coin
        domains.go:118: 
            	Error Trace:	domains.go:118
            	Error:      	Not equal: 

make functional 

    --- PASS: TestApis//v2/staking/delegations (0.00s)
    reporter.go:23: 
        	Error Trace:	reporter.go:23
        	            				chain.go:21
        	            				response.go:585
        	            				response.go:151
        	            				http.go:57
        	            				http.go:92
        	            				asm_amd64.s:1357
        	Error:      	
        	            	expected status equal to:
        	            	 "200 OK"
        	            	
        	            	but got:
        	            	 "400 Bad Request"
        	Test:       	TestApis

**Comment by EnoRage** [comment]
> @EnoRage
> I can only run make test without any issue.
> make integration and make functional all failed even if I use the origin/master branch without any modification. Do I need to source any script before running tests ?
> 
> which test case will trigger this error ? I am trying to duplicate and find out the reason...
> 
> http://localhost:8420/v2/harmony/staking/delegations/one1e4mr7tp0a76wnhv9xd0wzentdjnjnsh3fwzgfv
> 
> {
> "code": 500,
> "error": {
> "error": "json: cannot unmarshal array into Go value of type harmony.Delegations: json-rpc GetObject Unmarshal error: Unable to fetch delegations list",
> "meta": {
> "obj": {
> "result": null
> },
> "string": "[]"
> }
> }
> }
> 
> make integration failed:
> 
> --- FAIL: Test/test_batch_with_invalid_coin (0.00s)
> domains.go:117:
> Error Trace: domains.go:117
> Error: Expected nil, but got: &errors.Error{Err:(*errors.errorString)(0xc000c3a760), Type:0x0, meta:map[string]interface {}{"coins":[]uint64{0x2c, 0x3c}, "name":"vitalik.eth"}, stack:[]string(nil)}
> Test: Test/test_batch_with_invalid_coin
> domains.go:118:
> Error Trace: domains.go:118
> Error: Not equal:
> 
> make functional
> 
> ```
> --- PASS: TestApis//v2/staking/delegations (0.00s)
> reporter.go:23: 
>     	Error Trace:	reporter.go:23
>     	            				chain.go:21
>     	            				response.go:585
>     	            				response.go:151
>     	            				http.go:57
>     	            				http.go:92
>     	            				asm_amd64.s:1357
>     	Error:      	
>     	            	expected status equal to:
>     	            	 "200 OK"
>     	            	
>     	            	but got:
>     	            	 "400 Bad Request"
>     	Test:       	TestApis
> ```

Currently it is not possible to run all the tests without production endpoints that are not public. We are working to mock all of the data from them. 

I reccomend just to fix issue with
http://localhost:8420/v2/harmony/staking/delegations/one1e4mr7tp0a76wnhv9xd0wzentdjnjnsh3fwzgfv

To test it locally just:

cd cmd/platform_api

ATLAS_PLATFORM=harmony go run main.go

Try GET request that was mentioned before.

You can also debug only platform_api binary. I saw, that your client implementation make a RPC request that does not have exist. See before. Also, can you please reopen the PR, because I cannot do it on my own?

**Pull request merged**